java/flight/src/main/java/org/apache/arrow/flight/FlightServer.java

```diff
@@ -90,20 +86,86 @@ public void close() throws InterruptedException {
     }
   }
 
-  public interface OutputFlight {
-    void sendData(int count);
+  public static Builder builder(BufferAllocator allocator, Location location, FlightProducer producer) {
+    return new Builder(allocator, location, producer);
+  }
 
-    void done();
+  public static final class Builder {
 
-    void fail(Throwable t);
-  }
+    private final BufferAllocator allocator;
+    private final Location location;
+    private final FlightProducer producer;
+    private ServerAuthHandler authHandler = ServerAuthHandler.NO_OP;
 
-  public interface FlightServerHandler {
+    private InputStream certChain;
+    private InputStream key;
 
-    public FlightInfo getFlightInfo(String descriptor) throws Exception;
+    Builder(BufferAllocator allocator, Location location, FlightProducer producer) {
+      this.allocator = allocator;
+      this.location = location;
+      this.producer = producer;
+    }
 
-    public OutputFlight setupFlight(VectorSchemaRoot root);
+    public FlightServer build() {
+      final NettyServerBuilder builder;
+      switch (location.getUri().getScheme()) {
+        case LocationSchemes.GRPC_DOMAIN_SOCKET: {
+          // TODO: need reflection to check if domain sockets are available
+          throw new UnsupportedOperationException("Domain sockets are not available.");
+        }
+        case LocationSchemes.GRPC:
+        case LocationSchemes.GRPC_INSECURE: {
+          builder = NettyServerBuilder
+              .forAddress(new InetSocketAddress(location.getUri().getHost(), location.getUri().getPort()));
+          break;
+        }
+        case LocationSchemes.GRPC_TLS: {
+          if (certChain == null) {
+            throw new IllegalArgumentException("Must provide a certificate and key to serve gRPC over TLS");
+          }
+          builder = NettyServerBuilder
+              .forAddress(new InetSocketAddress(location.getUri().getHost(), location.getUri().getPort()));
+          break;
+        }
+        default:
+          throw new IllegalArgumentException("Scheme is not supported: " + location.getUri().getScheme());
+      }
+
+      if (certChain != null) {
+        builder.useTransportSecurity(certChain, key);
+      }
+
+      final Server server = builder
+          .executor(new ForkJoinPool())
```

Comment:
How about we make the executor configurable as part of this builder? Same with max message size below.

Reply:
Sure - I was going to do that in a followup, but I'll include that here.